### ISSUE TITLE
Generated hashCode() function not generating as expected

### ISSUE THREAD

**Issue by dalippa** [immutables/immutables#449]
I was stepping through our code and found this in our generated code:

``` java
@Override
  public int hashCode() {
    int h = 31;
    h = h * 17 + firstInstant.hashCode();
    h = h * 17 + secondInstant.hashCode();
    return h;
  }
```

This did not seem to be what I was expecting in hashCode(). I was expecting this instead:

``` java
@Override
  public int hashCode() {
    int h = 17;
    h = h * 31 + firstInstant.hashCode();
    h = h * 31 + secondInstant.hashCode();
    return h;
  }
```

I noted this looked odd since I remember seeing bit shifting to speed up hashCode generation since:

``` java
 31 * i == (i << 5) - i 
```

So two questions here:
1. Is the generated hashCode() working as designed here?
2. If the second implementation was what had been desired, are there plans for using bit-shifting for performance improvements?

**Comment by elucash** [comment]
Thank you for the suggestion! There are many recipes for hash code generation. For this one, the only requirement is that numeric multipliers should be prime numbers. If there will be enouph interest and performance gains could be confirmed we might change the implementation. Another possibility is to try to make algorithm parametrizable, not like you override the whole hashCode, but only specify formula/template for computing each value, but it would be harder/longer to implement. Most of the time performance problems could be solved by precomputing hashCode values: http://immutables.github.io/immutable.html#precomputed-hashcode

**Comment by tinkerware** [comment]
I have to say that this really looks like a typo that got baked in as a mistake. The generated code is straight out of the Effective Java item 9, except `17` and `31` have switched places. The original point about `31` being easily optimized by a compiler into bit shift is given as the motivating factor for the choice of that constant; that it is a prime is less important than the fact that it is odd, as described in the book.

**Comment by elucash** [comment]
I always thought that primes give better bit distribution, too naive probably..
Maybe we can switch 17 and 31 for the 2.4 version bump.
@tinkerware @dalippa are there constants that are any better?

**Comment by tinkerware** [comment]
> I always thought that primes give better bit distribution, too naive probably..
> Maybe we can switch 17 and 31 for the 2.4 version bump.
> @tinkerware @dalippa are there constants that are any better?

That's a good question. I did a little search, and for simple multiplicative hashes of the sort we use here, I found this set of guidelines from https://www.strchr.com/hash_functions:

> There are no exact rules for choosing the multiplier, only some heuristics:

>   - the multiplier should be large enough to accommodate most of the possible letters (e.g., 3 or 5 is too small);
>   - the multiplier should be fast to calculate with shifts and additions [e.g., 33 * hash can be calculated as (hash << 5) + hash];
>   - the multiplier should be odd for the reason explained above;
>   - prime numbers are good multipliers.

`33` seems to be favored by DJB (https://cr.yp.to/djb.html); the `bernstein` hash function has an initial value of `5381` and a multiplier of `33`. In hindsight, `17` is not too bad on how fast it is, since the compiler likely optimizes it to` h << 4 + h`, but there are better choices.

**Comment by elucash** [comment]
If we are to change hashcode for the next release, I like the idea of initial value such as `5381` and `33`. Maybe even make optimization explicit, it does not add any overhead in the code:

```java
public int hashCode() {
  int h = 5381;
  h += (h << 5) + Doubles.hashCode(red);
  h += (h << 5) + Doubles.hashCode(green);
  h += (h << 5) + Doubles.hashCode(blue);
  return h;
}
```

**Comment by tinkerware** [comment]
> h += (h << 5) + Doubles.hashCode(red);

That should be `h += (h << 5) + h + Doubles.hashCode(red);` and so on for the others.

**Comment by elucash** [comment]
I've used `+=` instead of `=`

```
h = h * 33 + perFieldCode
h = (h << 5) + h + perFieldCode
h += (h << 5) + perFieldCode
```

**Comment by tinkerware** [comment]
Ah, tricky. I like it when compilers make tricky code unnecessary :).

**Comment by elucash** [comment]
I'll have to jmh this some time before shipping

**Comment by tinkerware** [comment]
http://stackoverflow.com/questions/1579721/why-are-5381-and-33-so-important-in-the-djb2-algorithm has a good summary of how the constants in this hash function were picked.

**Comment by dalippa** [comment]
@elucash I liked your 5381 hashCode() function. It looks like it would be more straightforward.

I am using Immutables in production code at the moment, particularly where DateTime/Instant/ZonedDateTime objects are recreated frequently, so anything that speeds computation and ensures good distribution is welcome.

**Comment by elucash** [comment]
released in 2.4.0, thank you!
